crates/relayer/src/chain/ckb/utils.rs

```diff
@@ -0,0 +1,223 @@
+    let (packed_headers_mmr_root, packed_headers_mmr_proof) = {
+        let positions = (start_slot..=maximal_slot)
+            .into_iter()
+            .map(|slot| mmr::lib::leaf_index_to_pos(slot - minimal_slot))
+            .collect::<Vec<_>>();
+
+        let mmr = storage.chain_root_mmr(maximal_slot)?;
+
+        let headers_mmr_root = mmr.get_root().map_err(StorageError::from)?;
+        let headers_mmr_proof_items = mmr
+            .gen_proof(positions)
+            .map_err(StorageError::from)?
+            .proof_items()
+            .iter()
+            .map(Clone::clone)
+            .collect::<Vec<_>>();
+        let headers_mmr_proof = packed::MmrProof::new_builder()
+            .set(headers_mmr_proof_items)
+            .build();
+
+        (headers_mmr_root, headers_mmr_proof)
+    };
+
+    // Build the packed proof update.
+    let packed_proof_update = {
+        let updates_items = finalized_headers
+            .iter()
+            .map(|header| {
+                packed::FinalityUpdate::new_builder()
+                    .finalized_header(header.inner.pack())
+                    .build()
+            })
+            .collect::<Vec<_>>();
+        let updates = packed::FinalityUpdateVec::new_builder()
+            .set(updates_items)
+            .build();
+        packed::ProofUpdate::new_builder()
+            .new_headers_mmr_root(packed_headers_mmr_root)
+            .new_headers_mmr_proof(packed_headers_mmr_proof)
+            .updates(updates)
+            .build()
+    };
+
+    // Invoke verification from core::Client on packed_proof_update
+    let client = if let Some(client) = onchain_packed_client_opt {
+        client
+            .unpack()
+            .try_apply_packed_proof_update(packed_proof_update.as_reader())
+            .map_err(|_| Error::send_tx("failed to update header".to_owned()))?
+    } else {
+        EthLcClient::new_from_packed_proof_update(packed_proof_update.as_reader())
+            .map_err(|_| Error::send_tx("failed to create header".to_owned()))?
+    };
+
+    Ok((client.pack(), packed_proof_update))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::get_verified_packed_client_and_proof_update;
+    use super::EthHeader;
+    use super::EthUpdate;
+    use eth2_types::MainnetEthSpec;
+    use eyre::Result;
+    use ibc_relayer_storage::Storage;
+    use std::fs;
+    use std::path::Path;
+
+    fn load_updates_from_file(path: &str) -> Result<Vec<EthUpdate>> {
+        let headers_json = fs::read_to_string(path)?;
+        let headers: Vec<EthHeader> = serde_json::from_str(&headers_json)?;
+        Ok(headers
+            .into_iter()
+            .map(EthUpdate::from_finalized_header)
+            .collect::<Vec<_>>())
+    }
+
+    #[test]
+    fn test_get_verified_packed_client_and_proof_update() {
+        let path = Path::new("test_storage");
+        if path.is_dir() {
+            fs::remove_dir_all(path).unwrap();
+        }
```

Comment:
I prefer to [` tempfile::TempDir`].

With [` tempfile::TempDir`], the directory will be automatically deleted when it goes out of scope.

[` tempfile::TempDir`]: https://docs.rs/tempfile/latest/tempfile/struct.TempDir.html

Reply:
it's a great suggestion and also a good knowledge point!